addons/binding/org.openhab.binding.fronius/src/main/java/org/openhab/binding/fronius/handler/FroniusBaseThingHandler.java

```diff
@@ -0,0 +1,154 @@
+import org.openhab.binding.fronius.internal.api.ValueUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Basic Handler class for all Fronius services.
+ *
+ * @author Gerrit Beine - Initial contribution
+ * @author Thomas Rokohl - Refactoring to merge the concepts
+ */
+public abstract class FroniusBaseThingHandler extends BaseThingHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(FroniusBaseThingHandler.class);
+    private final String serviceDescription;
+    private FroniusBridgeHandler bridgeHandler;
+
+    public FroniusBaseThingHandler(Thing thing) {
+        super(thing);
+        serviceDescription = getDescription();
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command == RefreshType.REFRESH) {
+            updateChannels();
+        }
+    }
+
+    @Override
+    public void initialize() {
+        if (getFroniusBridgeHandler() == null) {
+            logger.debug("Initializing {} Service is only supported within a bridge", serviceDescription);
+            updateStatus(ThingStatus.OFFLINE);
+            return;
+        }
+        logger.debug("Initializing {} Service", serviceDescription);
+        getFroniusBridgeHandler().registerService(this);
+    }
+
+    private synchronized FroniusBridgeHandler getFroniusBridgeHandler() {
+        if (this.bridgeHandler == null) {
+            Bridge bridge = getBridge();
+            if (bridge == null) {
+                return null;
+            }
+            ThingHandler handler = bridge.getHandler();
+            if (handler instanceof FroniusBridgeHandler) {
+                this.bridgeHandler = (FroniusBridgeHandler) handler;
+            } else {
+                return null;
+            }
+        }
+        return this.bridgeHandler;
+    }
+
+    /**
+     * Update all Channels
+     */
+    protected void updateChannels() {
+        for (Channel channel : getThing().getChannels()) {
+            updateChannel(channel.getUID().getId());
+        }
+    }
+
+    /**
+     * Update the channel from the last data
+     *
+     * @param channelId the id identifying the channel to be updated
+     */
+    protected void updateChannel(String channelId) {
+        if (!isLinked(channelId)) {
+            return;
+        }
+        Object value;
+        value = getValue(channelId);
+
+        State state = null;
+        if (value == null) {
+            state = UnDefType.UNDEF;
+        } else if (value instanceof Calendar) {
+            state = new StringType(((Calendar) value).toString());
+        } else if (value instanceof BigDecimal) {
+            state = new DecimalType((BigDecimal) value);
+        } else if (value instanceof Integer) {
+            state = new DecimalType(BigDecimal.valueOf(((Integer) value).longValue()));
+        } else if (value instanceof String) {
+            state = new StringType(value.toString());

```

Comment:
Do we have any String channel at all? If not, better do not create a state in this case as you will send an invalid update below.